File: Data Structures & Algorithms/maximum-subarray/submission-39.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxSubArray(nums) {
        const dfs = (l, r) => {
            if (l > r) {
                return -Infinity;
            }
            let m = (l + r) >> 1;
            let leftSum = 0, rightSum = 0, curSum = 0;
            for (let i = m - 1; i >= l; i--) {
                curSum += nums[i];
                leftSum = Math.max(leftSum, curSum);
            }

            curSum = 0;
            for (let i = m + 1; i <= r; i++) {
                curSum += nums[i];
                rightSum = Math.max(rightSum, curSum);
            }
            return Math.max(dfs(l, m - 1), 
                        Math.max(dfs(m + 1, r), 
                            leftSum + nums[m] + rightSum));
        }
    }
}

File: Data Structures & Algorithms/regular-expression-matching/submission-39.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        int m = s.Length, n = p.Length;
        return Dfs(0, 0, s, p, m, n);
    }

    private bool Dfs(int i, int j, string s, string p, int m, int n) {
        if (j == n) {
            return i == m;
        }

        bool match = i < m && (s[i] == p[j] || p[j] == '.');
        if (j + 1 < n && p[j + 1] == '*') {
            return Dfs(i, j + 2, s, p, m, n) || 
                   (match && Dfs(i + 1, j, s, p, m, n));
        }

        if (match) {
            return Dfs(i + 1, j + 1, s, p, m, n);
        }

        return false;
    }
}